Reworking docs, quickstart, and zenml go

htahir1 opened the pull request:
# READ THIS FIRST

This PR was intended to be a docs change but has now morphed into misc code changes and a skeleton of docs. Please only review the code changes and ignore the docs for now. ill merge it in for the release and then we can do docs in another branch.

Please also ignore the `examples/quickstart` changes. @fa9r is taking care of this in another branch

## Describe changes
Im keeping this as a running doc of things im changing:

- Reworked quickstart to now have more integrations 
- Made dash visualizer work within a notebook
- zenml go now clones the entire zenml repository to a temp path and copies the quickstart to the current working directory
- zenml go now sends an analytics event
- zenml clean now preserves user_metadata
- Fixed error with importing the GCP Artifact Store
- Changed telemetry to now all go through the ZenStores rather than Repository. This has required extensive changes to the codebase but sets us up well.

# Docs
- Added best practices
- Introduces a new recommended repo structure [IMPORTANT:WE NEED TO UPDATE OUR REPOSITORIES TO FOLLOW THIS STRUCTURE AT SOME POINT]
- Reworked advanced guide
- Added FAQ's
- Refactored Repo and Global Config to various parts of the doc
- Added `zenml explain` in the CLI docs


## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [x] I have added tests to cover my changes.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Other (add details above)

strickvl commented on .gitignore:
```diff
@@ -176,4 +176,8 @@ zenml_examples/
+!src/zenml/integrations/wandb
+
+
+# some examples have yamls, dont ignore them
```
```suggestion
# some examples have yamls; don't ignore them
```

strickvl commented on docs/book/reference/glossary.md:
```diff
@@ -130,6 +130,23 @@ configuration lives, e.g., the active
 
+## Runner Scripts
+
+A runner script is a python file, usually called `run.py` and located at the root of a 
```
```suggestion
A runner script is a Python file, usually called `run.py` and located at the root of a 
```

strickvl commented on docs/book/support/best-practices.md:
````diff
@@ -0,0 +1,56 @@
+└── .zen
+```
+
+# Best Practices
+
+# Best Practices and Tips
````
Will these be explained? Ignore if this is just a draft / beginning of this page.

strickvl commented on src/zenml/cli/__init__.py:
````diff
@@ -47,6 +47,16 @@
+zenml metadata-store explain
+```
+
+This will give you an explanation of that concept in more detail/
````
```suggestion
This will give you an explanation of that concept in more detail.
```

strickvl commented on src/zenml/cli/base.py:
```diff
@@ -213,8 +234,9 @@ def go() -> None:
-    """Ask the user to give their email address"""
+def _prompt_email(gc: GlobalConfiguration) -> bool:
+    """Ask the user to give their email address. Returns
+    True if email is given, else returns False."""
```
```suggestion
    True if email is given, else False."""
```

strickvl commented on src/zenml/zen_stores/rest_zen_store.py:
```diff
@@ -310,21 +301,12 @@ def update_stack(self, name: str, stack: StackWrapper) -> Dict[str, str]:
-        Returns:
-            metadata dict for telemetry or logging.
-
         Raises:
             ValueError: If a dict is not returned from the API.
         """
```
If you've deleted the raising of a `ValueError`, then this should be deleted as well.

Not quite sure why you deleted that check, though.

strickvl reviewed (commented):
I ignored the `docs` parts, but the rest looks good to me aside from the query I had below...

stefannica commented on src/zenml/zen_stores/base_zen_store.py:
```diff
@@ -149,8 +152,10 @@ def stack_configurations(self) -> Dict[str, Dict[StackComponentType, str]]:
+
     @abstractmethod
-    def register_stack_component(
+    def _register_stack_component(
```
Ah, now that I've gone through the entire PR, I understand why you had to do this.

In cases like this, the recommended object-oriented design approach is not to alter the base class and all inheriting classes, but to implement a ["proxy"](https://refactoring.guru/design-patterns/proxy/python/example) instead: a new ZenStore class also inheriting from BaseZenStore that wraps around around an existing BaseZenStore instance supplied during construction, and intermediates all method calls to implement analytics:

```
class AnalyticsEnabledZenStore(BaseZenStore):
    def __init__(
        self,
        zen_store_type: Type[BaseZenStore],
        *args: Any,
        **kwargs: Any,
    ) -> "AnalyticsEnabledZenStore":
        self._wrapped_store = zen_store_type(*args, **kwargs)

    def initialize(
        self,
        *args: Any,
        **kwargs: Any,
    ) -> "AnalyticsEnabledZenStore":
        self._wrapped_store.initialize(*args, **kwargs)
        return self

...

    def register_stack_component(
        self,
        component: ComponentWrapper,
    ) -> None:
        analytics_metadata = {
            "type": component.type.value,
            "flavor": component.flavor,
        }
        self._track_event(
            AnalyticsEvent.REGISTERED_STACK_COMPONENT,
            metadata=analytics_metadata,
        )
        return self._wrapped_store.register_stack_component(component)
```

stefannica reviewed (changes requested):
I mainly reviewed the ZenStore and Repository changes. If you have the time and the energy for it, I would really recommend what I think is a better approach of modifying an existing class to add more concerns to it that are unrelated to its current set of responsibilities, one that doesn't require you to change the existing base class and its implementations.

AlexejPenner commented on src/zenml/zen_stores/base_zen_store.py:
```diff
@@ -55,6 +56,7 @@ def initialize(
             url: The URL of the store.
             skip_default_registrations: If `True`, the creation of the default
                 stack and user will be skipped.
+            track_analytics: If `False`, then we do not track events.
```
same as above

AlexejPenner reviewed (changes requested):
Small docstring nits, would also be interested to see your response to Stefans point. Was also wondering about the private methods.

htahir1 commented:
@AlexejPenner @stefannica I have addressed both your comments. 

The one I have neglected is the big refactor suggestion, which while being great, both @stefannica and I have decided is better left for later. Added a TODO to reflect this!

stefannica reviewed (approved):
LGTM !

strickvl commented on docs/book/misc/best-practices.md:
```diff
@@ -0,0 +1,109 @@
+
+This ensures that loading the pipeline from elsewhere does not also run it.
+
+#### Don't use the same metadata stores across  multiple artifact stores
```
```suggestion
#### Don't use the same metadata stores across multiple artifact stores
```

strickvl commented on docs/book/misc/best-practices.md:
```diff
@@ -0,0 +1,109 @@
+
+#### Use unique pipeline names across projects, especially if used with the same metadata store
+
+Pipeline names are their unique identifiers, as such using the same name for
```
```suggestion
Pipeline names are their unique identifiers, so using the same name for
```

strickvl commented on docs/book/misc/best-practices.md:
```diff
@@ -0,0 +1,109 @@
+
+#### Check which integrations are required for registering a stack component 
+
+You can do so by running zenml flavor list and installing the integration(s) 
```
```suggestion
You can do so by running `zenml flavor list` and installing the integration(s) 
```

strickvl commented on docs/book/misc/best-practices.md:
```diff
@@ -0,0 +1,109 @@
+
+#### Initialize the zenml repository in the root of the source code tree of a project, even if it's optional
+
+This will set the zenml project root for the project and ...
```
TO BE ADDED

strickvl commented on docs/book/misc/contributing.md:
```diff
@@ -0,0 +1,23 @@
+
+## 🆘 Where to get help
+
+First point of call should be [our Slack group](https://zenml.io/slack-invite/). Ask your questions about bugs or specific use cases and someone from the core team will respond.
```
```suggestion
The first point of call should be [our Slack group](https://zenml.io/slack-invite/). Ask your questions about bugs or specific use cases and someone from the core team will respond.
```

strickvl commented on docs/book/misc/contributing.md:
```diff
@@ -0,0 +1,23 @@
+
+
+
+ZenML is distributed under the terms of the Apache License Version 2.0. A complete version of the license is available in the [LICENSE.md](https://github.com/zenml-io/zenml/blob/main/LICENSE.md) in this repository. Any contribution made to this project will be licensed under the Apache License Version 2.0.
```
```suggestion
```

Duplicated below

strickvl commented on docs/book/misc/faq.md:
```diff
@@ -0,0 +1,68 @@
+---
+description: We answer the most frequently asked questions about ZenML here
```
```suggestion
description: We answer the most frequently asked questions about ZenML here.
```

strickvl commented on docs/book/misc/faq.md:
```diff
@@ -0,0 +1,68 @@
+
+# FAQ
+
+#### How does ZenML compare to orchestrators (Airflow, Kubeflow, Flyte, etc.)
```
```suggestion
#### How does ZenML compare to orchestrators (Airflow, Kubeflow, Flyte, etc.)?
```

strickvl commented on docs/book/misc/faq.md:
```diff
@@ -0,0 +1,68 @@
+
+An orchestrator in MLOps is the system component that is responsible for executing and managing the execution of a ML pipeline. ZenML pipelines can be run on multiple orchestration systems. There are [standard orchestrators](../extending-zenml/orchestrator.md) that ZenML supports out-of-the-box, but you are encourage to write your own [orchestrator](../extending-zenml/orchestrator.md) in order to ...
+
+#### Can I use tool X? How does Y tool integrate with ZenML?
```
```suggestion
#### Can I use tool X? How does tool Y integrate with ZenML?
```

strickvl commented on docs/book/misc/faq.md:
```diff
@@ -0,0 +1,68 @@
+
+#### Can I use tool X? How does Y tool integrate with ZenML?
+
+Take a look at our [examples](https://github.com/zenml-io/zenml/tree/main/examples) directory, which showcases detailed example for each integration ZenML supports out-of-the-box.
```
```suggestion
Take a look at our [examples](https://github.com/zenml-io/zenml/tree/main/examples) directory, which showcases detailed examples for each integration that ZenML supports out-of-the-box.
```

strickvl commented on docs/book/misc/faq.md:
```diff
@@ -0,0 +1,68 @@
+
+Take a look at our [examples](https://github.com/zenml-io/zenml/tree/main/examples) directory, which showcases detailed example for each integration ZenML supports out-of-the-box.
+
+The ZenML team and community is constantly working to include more tools and integrations to the above list (check the [roadmap](https://zenml.io/roadmap) for more details). You can [upvote features](https://zenml.io/discussion) you'd like and add more ideas on the roadmap.
```
```suggestion
The ZenML team and community is constantly working to include more tools and integrations to the above list (check out the [roadmap](https://zenml.io/roadmap) for more details). You can [upvote features](https://zenml.io/discussion) you'd like and add your ideas to the roadmap.
```

strickvl commented on docs/book/misc/faq.md:
```diff
@@ -0,0 +1,68 @@
+
+#### Why did you build ZenML?
+
+We built it because we scratched our own itch while deploying multiple ML models in production for the last 3 years. Our team struggled to find a simple yet production-ready solution whilst developing large scale ML pipelines, and built a solution for it that we are now proud to share with all of you!
```
```suggestion
We built it because we scratched our own itch while deploying multiple machine learning models in production for the past three years. Our team struggled to find a simple yet production-ready solution whilst developing large-scale ML pipelines, and built a solution for it that we are now proud to share with all of you!
```

strickvl commented on docs/book/misc/faq.md:
```diff
@@ -0,0 +1,68 @@
+
+#### How can I contribute?
+
+We would love to develop ZenML together with our community! Best way to get
```
```suggestion
We would love to develop ZenML together with our community! The best way to get
```

strickvl commented on docs/book/misc/faq.md:
```diff
@@ -0,0 +1,68 @@
+
+#### Why should I use ZenML?
+
+ZenML pipelines are designed to be written early on the development lifecycle. Data scientists can explore their pipelines as they develop towards production, switching stacks from local to cloud deployments with ease. You can read more about why we started building ZenML on our blog. By using ZenML in the early stages of your project, you get the following benefits:
```
```suggestion
ZenML pipelines are designed to be written early on the development lifecycle. Data scientists can explore their pipelines as they develop towards production, switching stacks from local to cloud deployments with ease. You can read more about why we started building ZenML [on our blog](https://blog.zenml.io/). By using ZenML in the early stages of your project, you get the following benefits:
```

strickvl commented on examples/quickstart/README.md:
```diff
@@ -1,59 +1,92 @@
-### 📄 Prerequisites 
-In order to run this example, you need to install and initialize ZenML:
+This quickstart aims to give you a small illustration of what ZenML can do. To do so, we will:
+- Train a model on the Digits dataset, evaluate it, deploy it, and embed it in an inference pipeline,
```
```suggestion
- Train a model on the MNIST digits dataset, evaluate it, deploy it, and embed it in an inference pipeline
```

strickvl commented on examples/quickstart/README.md:
```diff
@@ -1,59 +1,92 @@
-In order to run this example, you need to install and initialize ZenML:
+This quickstart aims to give you a small illustration of what ZenML can do. To do so, we will:
+- Train a model on the Digits dataset, evaluate it, deploy it, and embed it in an inference pipeline,
+- Automatically version, track, and cache data, models, and other artifacts,
```
```suggestion
- Automatically version, track, and cache data, models, and other artifacts
```

strickvl commented on examples/quickstart/README.md:
```diff
@@ -1,59 +1,92 @@
+This quickstart aims to give you a small illustration of what ZenML can do. To do so, we will:
+- Train a model on the Digits dataset, evaluate it, deploy it, and embed it in an inference pipeline,
+- Automatically version, track, and cache data, models, and other artifacts,
+- Track model hyperparameters and metrics in an experiment tracking tool,
```
```suggestion
- Track model hyperparameters and metrics in an experiment tracking tool
```

strickvl commented on examples/quickstart/README.md:
```diff
@@ -1,59 +1,92 @@
+- Train a model on the Digits dataset, evaluate it, deploy it, and embed it in an inference pipeline,
+- Automatically version, track, and cache data, models, and other artifacts,
+- Track model hyperparameters and metrics in an experiment tracking tool,
+- Measure and visualize train-test skew, training-serving skew, and data drift.
```
```suggestion
- Measure and visualize train-test skew, training-serving skew, and data drift
```

strickvl commented on examples/quickstart/README.md:
```diff
@@ -1,59 +1,92 @@
+- Track model hyperparameters and metrics in an experiment tracking tool,
+- Measure and visualize train-test skew, training-serving skew, and data drift.
+
+**New to MLOps?** Then you might want to start with our [**ZenBytes**](https://github.com/zenml-io/zenbytes) lesson series instead, where we cover each MLOps concept in much more detail. This quickstart assumes you are already familiar with basic MLOps issues and just want to learn how to approach them with ZenML.
```
```suggestion
**New to MLOps?** You might want to start with our [**ZenBytes**](https://github.com/zenml-io/zenbytes) lesson series instead where we cover each MLOps concept in much more detail. This series assumes you are already familiar with basic MLOps issues and just want to learn how to approach them with ZenML.
```

strickvl reviewed (commented):
Small things only.
